# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Цель исследования: проанализировать базу данных крупного сервиса для чтения книг по подписке 

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
!pip install psycopg2-binary

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Обзор данных

In [4]:
query_books = ''' SELECT * FROM books; ''' 
books = pd.io.sql.read_sql(query_books, con = engine)
books.head(5)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [6]:
query_authors = ''' SELECT * FROM authors; ''' 
authors = pd.io.sql.read_sql(query_authors, con = engine)
authors.head(5)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [8]:
query_publishers = ''' SELECT * FROM publishers; ''' 
publishers = pd.io.sql.read_sql(query_publishers, con = engine)
publishers.head(5)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [10]:
query_ratings = ''' SELECT * FROM ratings; ''' 
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
ratings.head(5)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [12]:
query_reviews = ''' SELECT * FROM reviews; ''' 
reviews = pd.io.sql.read_sql(query_reviews, con = engine)
reviews.head(5)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


### Задания

**Задание 1**

Посчитайте, сколько книг вышло после 1 января 2000 года

In [14]:
book_query_1 = '''SELECT COUNT(*)
                  FROM books
                  WHERE publication_date > '2000-01-01';'''
request_1 = pd.io.sql.read_sql(book_query_1, con = engine)
request_1

count
0    819

После 1 января 2000 года вышло 819 книг.

**Задание 2**

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [15]:
book_query_2 = '''SELECT b.book_id,
b.title,
r.avg_rating,
rv.review_count
FROM books as b
LEFT JOIN 
    (SELECT book_id,
     AVG(rating) as avg_rating
     FROM ratings
     GROUP BY book_id) as r on r.book_id=b.book_id
LEFT JOIN 
    (SELECT book_id,
     COUNT(review_id) as review_count
     FROM reviews 
     GROUP BY book_id) as rv on rv.book_id=b.book_id'''

request_2 = pd.io.sql.read_sql(book_query_2, con = engine)
request_2

book_id                                              title  avg_rating  \
0        652          The Body in the Library (Miss Marple  #3)    4.500000   
1        273                                          Galápagos    4.500000   
2         51                           A Tree Grows in Brooklyn    4.250000   
3        951  Undaunted Courage: The Pioneering First Missio...    4.000000   
4        839                                        The Prophet    4.285714   
..       ...                                                ...         ...   
995       64                                Alice in Wonderland    4.230769   
996       55          A Woman of Substance (Emma Harte Saga #1)    5.000000   
997      148                                          Christine    3.428571   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)    3.500000   
999      828                           The Plot Against America    3.000000   

     review_count  
0             2.0  
1             2.0  
2             5.0  
3             2.0  
4             4.0  
..            ...  
995           4.0  
996           2.0  
997           3.0  
998           2.0  
999           2.0  

[1000 rows x 4 columns]

**Задание 3**

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [16]:
book_query_3 = '''SELECT x.publisher,
x.count_book
FROM 
(SELECT b.publisher_id,
p.publisher,
COUNT(book_id) as count_book
FROM books as b
LEFT JOIN publishers as p on p.publisher_id=b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher,
b.publisher_id
ORDER BY count_book DESC) as x
LIMIT 1'''

request_3 = pd.io.sql.read_sql(book_query_3, con = engine)
request_3

publisher  count_book
0  Penguin Books          42

Издательство Penguin Books выпустило наибольшее число книг - 42 шт. - толще 50 страниц.

**Задание 4**

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [20]:
book_query_4 = '''SELECT x.author,
AVG(x.avg_rating)
FROM 
(SELECT a.author,
 AVG(r.rating) as avg_rating,
 COUNT(r.rating_id) as count_rating
 FROM books as b
 LEFT JOIN authors as a on b.author_id=a.author_id
 LEFT JOIN ratings as r on b.book_id=r.book_id
 GROUP BY b.book_id, a.author
 HAVING COUNT(r.rating_id) >= 50) as x
GROUP BY x.author
ORDER BY AVG(x.avg_rating) DESC
LIMIT 1'''

request_4 = pd.io.sql.read_sql(book_query_4, con = engine)
request_4

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

 J.K. Rowling/Mary GrandPré - авторы с самой высокой средней оценкой книг - 4.3

**Задание 5**

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [18]:
book_query_5 ='''SELECT AVG(x.count_review)
FROM 
(SELECT COUNT(*) as count_review
 FROM reviews
 WHERE username IN (SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(*) > 50)
 GROUP BY username) as x'''

request_5 = pd.io.sql.read_sql(book_query_5, con = engine)
request_5

avg
0  24.333333

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.3

**Итоговый вывод:**

Во всех таблицах отсутствуют пропуски, типы данных таже не пришлось менять.

После анализа данных можно сделать следующие выводы:
- После 1 января 2000 года вышло 819 книг.
- Издательство Penguin Books выпустило наибольшее число книг - 42 шт. - толще 50 страниц.
- J.K. Rowling/Mary GrandPré - авторы с самой высокой средней оценкой книг - 4.3
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.3

+ Для каждой книги была подсчитана средняя оценка и количество обзоров.

Данное исследование должно помочь компании в дальнейшем улучшении сервиса и реализации различных бизнес-проектов.